lib/Conversion/TorchToLinalg/TorchToLinalg.cpp

```diff
@@ -3394,6 +3394,87 @@ class ConvertAtenNumelOp : public OpConversionPattern<AtenNumelOp> {
 };
 } // namespace
 
+namespace {
+class ConvertAtenIndexSelectOp : public OpConversionPattern<AtenIndexSelectOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenIndexSelectOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+
+    Location loc = op.getLoc();
+    Value input = adaptor.self();
+    Value indices = adaptor.index();
+    RankedTensorType inputType = input.getType().cast<RankedTensorType>();
+    RankedTensorType indicesType = indices.getType().cast<RankedTensorType>();
+    RankedTensorType resultType = getTypeConverter()
+                                      ->convertType(op->getResult(0).getType())
+                                      .cast<RankedTensorType>();
+    Type elementType = resultType.getElementType();
+    unsigned inputRank = inputType.getRank();
+
+    int64_t dimInt;
+    if (!matchPattern(op.dim(), m_TorchConstantInt(&dimInt)))
+      return op->emitError("unimplemented: dim is not constant");
+
+    Value initTensor;
+    if (inputType.hasStaticShape() && indicesType.hasStaticShape()) {
+      SmallVector<int64_t> resultShape;
+      for (auto size : inputType.getShape())
+        resultShape.push_back(size);
+      resultShape[dimInt] = indicesType.getShape()[0];
+      initTensor =
+          rewriter.create<linalg::InitTensorOp>(loc, resultShape, elementType);
+    } else {
+      SmallVector<Value> resultShape = getTensorSizes(rewriter, loc, input);
+      resultShape[dimInt] = getTensorSizes(rewriter, loc, indices)[0];
+      initTensor =
+          rewriter.create<linalg::InitTensorOp>(loc, resultShape, elementType);
+    }
+
+    SmallVector<AffineExpr> inputExpr, indicesExpr, resultExpr;
+    SmallVector<StringRef> iteratorTypes;
+
+    for (unsigned i = 0; i < inputRank; i++) {
+      inputExpr.push_back(rewriter.getAffineDimExpr(i));
+      resultExpr.push_back(rewriter.getAffineDimExpr(i));
+      iteratorTypes.push_back(getParallelIteratorTypeName());
+    }
+
+    resultExpr[dimInt] = rewriter.getAffineDimExpr(inputRank);
+    indicesExpr.push_back(rewriter.getAffineDimExpr(inputRank));
+
+    auto indexingMaps =
+        AffineMap::inferFromExprList({inputExpr, indicesExpr, resultExpr});
+    iteratorTypes.insert(iteratorTypes.end(), "parallel");
+
+    Value finalRes =
+        rewriter
+            .create<linalg::GenericOp>(
+                loc, resultType, ValueRange{input, indices}, initTensor,
+                /*indexingMaps=*/indexingMaps,
+                /*iteratorTypes=*/iteratorTypes,
+                [&](OpBuilder &b, Location loc, ValueRange args) {
+                  Value indexInput =
+                      rewriter.create<linalg::IndexOp>(loc, dimInt);
```

Comment:
This payload doesn't seem right to me. index_select is basically a "gather", so I would expect the payload to contain a tensor.extract op.

Reply:
Here we have `args[0]`, an element of the `input` tensor, and we are using it. That's why we didn't require an extract tensor op.

Reply:
I think so with `tensor_extract` op you have to potential to remove the last `iter_var`. So your algorithm would look something like: 
``` 
for i in range(input.size[0]): 
       for j in range (index.size[0]):
           for k in range(input.size[2]):
               index_val = index[j]
               output[i][j][k] = tensor.extract(input, i, index_val, j)
```
Remember, for `tensor.extract` op the accesses don't have to be affine whereas for the payload args it has to be affine.